# `데이터 전처리(Data Preprocessing)`
><b>"Garbage In, Garbage Out"</b>&nbsp;&nbsp;:&nbsp;&nbsp;아무리 뛰어난 AI 알고리즘이라도 <b>지저분한 데이터(Garbage)</b>를 넣으면 <b>지저분한 결과(Garbage)</b>만 나온다. <br>
> 데이터 전처리는 데이터 분석 및 머신러닝 프로젝트 전체 시간의 <b>70~80%</b>를 차지하는 가장 핵심적인 파이프라인이다.

#### &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; `데이터 전처리` = 원본 데이터(Raw Data)를 ➡️ 분석 가능한 형태의 정제된 데이터로 가공하는 과정 

### 🛠️ 핵심적 전처리 기술

<details>
<summary>🔍 <b>데이터 정제 (Data Cleaning)</b></summary>

1. 결측치(Missing Value) 처리
2. 이상치(Outlier) 처리

</details>

<details>
<summary>🧮 <b>데이터 변환 및 인코딩 (Transformation & Encoding)</b> </summary>

1. 범주형 데이터 인코딩 (Categorical Encoding) — 라벨 인코딩 (Label Encoding), 원-핫 인코딩 (One-Hot Encoding) <br>
2. 수치형 데이터 스케일링 (Feature Scaling) — (스케일링 기법들:) StandardScaler, MinMaxScaler, RobustScaler, Log Transformation

</details>

<details>
<summary>⬇️ <b>차원 축소 및 불균형 해소</b> </summary>

1. 차원 축소 (Dimensionality Reduction) — 변수(Feature)가 너무 많으면 과적합(Overfitting) 및 연산 속도 저하가 발생하기 떄문에, 변수를 줄임.<details><summary>　</summary>
└ Feature Selection: 상관관계 분석, 중요도 스코어 등을 통해 유용한 변수만 추려냄. <br>
└ Feature Extraction (PCA 등): 기존 변수들을 조합하여 핵심 정보(분산)를 보존하는 새로운 축으로 압축.</details>

3. 데이터 불균형 (Class Imbalance) 처리 — 분류 문제에서 특정 클래스의 비율이 극단적으로 적을 때 (예: 암 진단 1% vs 정상 99%) 적용함. <br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&rarr; Undersampling/Oversampling
</details>

&nbsp;&nbsp;&nbsp;🧑‍🤝‍🧑<b>Merge (데이터 결합)</b><br>
&nbsp;&nbsp;&nbsp;💱<b>컬럼명 변경 (Column Renaming) 및 데이터 타입 변경 (Type Casting)</b><br>
&nbsp;&nbsp;&nbsp;✂️<b>데이터 분할 (Data Splitting)</b>

<br><br><br>
# 1. `결측치(Missing Value) 탐색과 처리`
> 결측치: 값이 비어있는 것

In [2]:
# ================================== 결측치 점검 코드 예제 ==================================
import pandas as pd
import numpy as np

#pd.read_csv() 함수를 사용해 파일을 불러와 df 변수에 저장
df = pd.read_csv("D:\[실습 파일] 이론 데이터셋\Part3 House Price.csv")

# 각 컬럼에 포함된 결측치의 총 개수를 계산해 missing_counts 변수에 저장 (pandas series 형태로 결과물이 저장됨)
missing_counts = df.isnull().sum()

# 결측치가 1개 이상 존재하는 컬럼만 필터랑한 뒤, 내림차순 정렬
print(missing_counts[missing_counts > 0].sort_values(ascending=False))

PoolQC          1453
MiscFeature     1406
Alley           1369
Fence           1179
MasVnrType       872
FireplaceQu      690
LotFrontage      259
GarageType        81
GarageYrBlt       81
GarageFinish      81
GarageQual        81
GarageCond        81
BsmtFinType2      38
BsmtExposure      38
BsmtFinType1      37
BsmtCond          37
BsmtQual          37
MasVnrArea         8
Electrical         1
dtype: int64


#### 📌 결측치의 규모, 결측치의 성격이 컬럼마다 다르기 떄문에, 각 컬럼에 맞는 최적의 처리 전략을 세워야 한다.<br><br><br>

## 🚩전략 1. 과감하게 제거하기(dropna)<br>
### 1. 행(Row) 제거
>결측치가 단 하나라도 포함된 행 전체를 제거함

In [3]:
# .shape 속성으로 df의 형태를 확인하여 "제거 전" 상태를 기록한다.
print("제거 전 데이터 형태:", df.shape)

# df.dropna()를 실행하여 결측치가 있는 모든 행을 제거한 결과를 df_drop_rows 데이터프레임에 저장한다.
df_drop_rows = df.dropna()

# 행 제거 후, dr_drop_rows의 형태를 다시 확인하여, 얼마나 많은 데이터가 사라졌는지 비교한다.
print("행 제거 후 데이터 형태:", df_drop_rows.shape) 

제거 전 데이터 형태: (1460, 81)
행 제거 후 데이터 형태: (0, 81)


#### 📌 dropna()의 강력함을 실행결과로 알 수 있다.<br><br> 1460개의 모든 행(주택정보)에 최소 하나 이상의 결측치가 존재했다는 의미다. 모든 데이터를 잃지않게, 실전에서 반드시 신중하게 사용하자!<br>

### 2. 열(column) 제거
> (1) dropna(axis=1) : axis=1 옵션을 사용하면 특정 열 전체를 제거할 수 있음.<br>
> (2) df.drop() 함수의 columns 인자에 제거하고 싶은 열 이름들의 리스트를 전달하여 명시적으로 해당 열을 삭제할 수 있다. <br>　　&rarr; 직접 선택한 특정 열만 제거할 수 있어 axis=1보다 안전함.

In [4]:
# 결측치가 너무 많은 "Alley", "PoolQC", "MiscFeature" 컬럼을 제거한다.
df_drop_cols = df.drop(columns=['Alley', 'PoolQC', 'MiscFeature'])

# 열 제거 후 데이터의 행과 열 개수를 확인한다.
print("제거 전 데이터 형태:", df.shape)
print("열 제거 후 데이터 형태:", df_drop_cols.shape)

제거 전 데이터 형태: (1460, 81)
열 제거 후 데이터 형태: (1460, 78)


<br><br><br>
## 🚩전략 2. 현명하게 채워넣기(fillna)
> 데이터를 무작정 제거하는 방법(dropna)보다는 비어 있는 칸에 적절한 값을 채워 넣는 대체(Imputation) 방법이 더 선호된다.<br>
> 어떤 값으로 채울지는 결측치의 '성격'에 따라 결정된다.

<br>

### 1. 시나리오 A: '없음(Absence)'을 의미하는 결측치 처리
> 결측치가 '존재하지 않음'이라는 명확한 정보를 담고 있을 때, 그 의미를 살리는 값으로 채워주는 전략이다.<br>
> `df[col].fillna('None')`　&rarr;　해당 열의 모든 결측치(NaN)를 'None'이라는 문자열로 채워, 그 새로운 버전의 열(Series)을 생성한다.<br>
> 　 　　　　　　　　　　　　　　 기존의 원본 데이터프레임(df)의 [col]열에 덮어씌어주면 끝이다.

In [8]:
# House Prices 데이터에서 GarageType(차고종류) 컬럼의 결측치는 '차고가 없음'을 의미한다. 
# 이 "존재하지 않음"이라는 정확한 정보와 의미를 살리기 위해,         범주형에는 df[col].fillna('None'),      숫자형에는 df[col].fillna(0)를 사용한다.

# 1) 치고와 관련된 범주형 컬럼들의 이름을 리스트로 묶어 정리한다.
garage_cols_cat = ['GarageType', 'GarageFinish', 'GarageQual', 'GarageCond']

# [범주형 데이터의 빈칸은 'None'이라는 문자열로 채운다.]
# 2) for 반복문을 통해 각 리스트의 컬럼을 하나씩 순회하며 결측치를 처리한다.
for col in garage_cols_cat:
    # .fillna() 의 결과를 다시 원래의 컬럼에 할당(저장)한다.
    df[col] = df[col].fillna('None')

# [숫자형 데이터의 빈칸은 0으로 채운다.]
# 3) 차고와 관련된 숫자형 컬럼들의 이름을 리스트로 묶어 정리한다.
garage_cols_num = ['GarageYrBlt', 'GarageArea', 'GarageCars']
for col in garage_cols_num:
    df[col] = df[col].fillna(0)

# 처리된 차고 관련 컬럼들의 결측치 개수를 다시 확인한다.
print("--- 결측치 처리 후 결과 ---")
check_cols = garage_cols_cat + garage_cols_num
print(df[check_cols].isnull().sum())   # 전체 데이터프레임(df)에서, check_cols 리스트에 들어있는 특정 컬럼들만 골라내,  isnull().sum() 하라.

--- 결측치 처리 후 결과 ---
GarageType      0
GarageFinish    0
GarageQual      0
GarageCond      0
GarageYrBlt     0
GarageArea      0
GarageCars      0
dtype: int64


<br>

### 2. 시나리오 B: 정보가 누락된 결측치 처리
> `정보가 단순히 누락된 경우`, 데이터의 전체적인 특징을 대표할 수 있는 <b>통계값</b>으로 채워준다.

<div style="border: 1px solid #e0e0e0; background-color: #f8f9fa; padding: 15px; border-radius: 8px;">
숫자형 컬럼의 경우  →  median() 사용  →  전체 값 중 순서상 가장 중앙에 위치하는 값으로 채운다.<br>
범주형 컬럼의 경우  →  mode() 사용  →  전체 값 중 가장 빈번하게 등장하는 값으로 채운다.
</div>

In [15]:
# [1] 'LotFrontage(부지와 연결된 도로 길이)'는 숫자형 데이터이므로, 전체 데이터의 중앙값으로 결측치를 채운다.

# median() 함수로 중앙값을 계산한다.
median_lot_frontage = df['LotFrontage'].median()

# 계산된 중앙값으로 채운 결과를 다시 'LotFrontage' 컬럼에 저장한다.
df['LotFrontage'] = df['LotFrontage'].fillna(median_lot_frontage)


# [2] 'Electrical(집에 설치된 전기 시스템의 종류)'은 범주형 데이터이므로, 가장 자주 나타나는 최빈값으로 결측치를 채운다.

# " .mode() 함수는 최빈값을 Series 형태로 반환하므로, 첫 번째 값[0]을 선택해야 한다. "
# => 값 하나만 나오는 게 아니라 아래처럼 순번(인덱스)이 붙은 상자(Series)에 담겨서 나온다.  그 목록에서 맨 위의 값 하나(=최빈값)를 쏙 꺼내기 위해 [0]을 붙여주는 것이다.
# # 예를 들어 최빈값이 'SBrkr'라고 할 때, .mode()의 결과:
#                                                      0    SBrkr
#                                                      Name: Electrical, dtype: object

mode_electrical = df['Electrical'].mode()[0]
df['Electrical'] = df['Electrical'].fillna(mode_electrical) # df['Electrical'] 컬럼 갖고와서, fillna()로 채워 결과물 만들고, 다시 원본 df['Electrical']에 저장

print("--- 결측치 처리 후 결과 ---")
print(df[['LotFrontage', 'Electrical']].isnull().sum())   # pandas에서 df[](대괄호 1개)에는 컬럼 이름 하나(문자열)만 들어갈 수 있다.
# df[[]]   에서 안쪽 []는 컬럼 이름들을 하나로 묶는 파이썬 List   =>   (df[]) 바깥쪽 대괄호는 그 리스트를 받아서 "리스트 안의 컬럼으로 데이터프레임 만들어" 명령함

--- 결측치 처리 후 결과 ---
LotFrontage    0
Electrical     0
dtype: int64


<br><br><br>
# 2. `범주형 데이터 인코딩: 컴퓨터가 문자를 이해하게 만드는 법`
> 인코딩(Encoding): 데이터에 포함된 모든 문자형 데이터를 모델이 이해할 수 있는 숫자 형태로 변환하는 과정

<br><br>

## 🚩방법 1. 라벨 인코딩(Label Encoding)
> 각 문자열 카테고리에 고유한 정수(0, 1, 2, ...)를 부여하는 방식　　　　(영어는 알파벳순으로, 한글은 가나다순으로)

In [16]:
# 인코딩 방식 예 )   ['S', 'C', 'Q']라는 세 종류의 승선항 데이터가 있다면, 각각 0, 1, 2라는 숫자 라벨로 변환되는 방식
# Label Encoding 작업에는 Scikit-learn의 LabelEncoder을 사용

# =============== 코드 예제 ===============
# 주택 데이터에서 품질을 나타내는 ExterQual(외부 마감 품질) 컬럼은, "TA(Typical/Average)", "Gd(Good)", "Ex(Excellent)", "Fa(Fair)" 네 가지 값으로 이루어져 있음

# 1) scikit-learn 라이브러리에서 LabelEncoder을 불러온다.
from sklearn.preprocessing import LabelEncoder

# 2) 인코딩을 수행할 LabelEncoder 객체를 생성한다.
le = LabelEncoder()

# 3) 원본 데이터의 ExterQual 컬럼 상위 5개를 확인한다.
print(" --- 인코딩 전 --- ")
print(df["ExterQual"].head())

# 4) .fit_transform() 을 사용해 ExterQual 컬럼을 Label Encoding 한다.
df['ExterQual_Encoded'] = le.fit_transform(df["ExterQual"])
# 'ExterQual' 컬럼의 고유한 값들('Ex', 'Fa', 'Gd', 'TA')이 무엇인지 학습(fit)하고,
# 이를 바탕으로 각 값을 숫자(0, 1, 2, 3)로 변환(Transform)한다.
# Label Encoding 후의 그 결과를 ExterQual_Encoded라는 새로운 컬럼에 저장한다.

# 5) 라벨 인코딩 후에 생성된 새로운 컬럼(ExterQual_Encoded)의 상위 5개를 확인한다.
print("\n --- 인코딩 후 --- ")
print(df[['ExterQual', 'ExterQual_Encoded']].head())

 --- 인코딩 전 --- 
0    Gd
1    TA
2    Gd
3    TA
4    Gd
Name: ExterQual, dtype: object

 --- 인코딩 후 --- 
  ExterQual  ExterQual_Encoded
0        Gd                  2
1        TA                  3
2        Gd                  2
3        TA                  3
4        Gd                  2


<br>

#### 📌 `실행결과`: 알파벳 순서에 따라 인코딩 됨.　 &rarr; 　Ex=0, Fa=1, Gd=2, TA=3<br> 
#### 📌 `Label Encoding의 단점`: Model이 숫자의 크기를 보고 <b>특성의 중요도나 순서로 오해</b>할 수 있다.<br>
TA(3)가 Gd(2)보다 더 중요하거나 높은 등급이라고 인식할 수 있다.<br>
만약 키테고리에 명확한 순서 (Ex. Bronze < Silver < Gold)가 있는 경우가 아니라면, <b>Label Encoding은 성능을 저하시키는 원인이 될 수 있다</b>.<br><br>

## 🚩방법 2. 원-핫 인코딩(One-Hot Encoding)
> 글자 형태의 범주형 데이터를 컴퓨터가 이해할 수 있도록 0과 1로 이루어진 여러 개의 열로 변환하는 기술<br>
> label encoding의 의도치 않은 순서ㆍ중요도 관계 문제를 해결함

#### 전체 카테고리 개수만큼 새로운 컬럼을 만들어낸다.
#### 그리고 해당되는 단 하나의 컬럼만 1 (Hot)로 켜고, 나머지는 0으로 꺼둔다.

In [19]:
# <데이터 소개> Neighborhood = 집이 위치한 '동네(지역/마을) 이름'을 나타내는 범주형 컬럼

# <사용할 함수 소개> One-Hot Encoding 작업에는 Pandas의 get_dummies() 함수를 사용


# 1. 원본 데이터의 Neighborhood 컬럼 상위 5개를 확인한다
print(" --- 인코딩 전 --- ")
print(df['Neighborhood'].head())

# 2. pd.get_dummies() : 원-핫 인코딩 하는 함수. 원하는 컬럼을 입력하면, 컬럼에 포함된 고유한 값 각각을 새로운 컬럼으로 만들어준다. *고유값(Unique values):겹치지 않는 값
neighborhood_dummies = pd.get_dummies(df['Neighborhood'], prefix='Nbrhd')
# Neighborhood 컬럼을 입력받아, 해당 컬럼의 모든 동네 이름(고유값)을 각각의 새로운 컬럼으로 변환한다.
# prefix='Nbrhd' 옵션: 새로 생성될 컬럼 이름 앞에 'Nbrhd_'라는 접두사를 붙여, 어떤 컬럼에서 파생되었는지 알기 쉽게 해준다.

# 3. 원-핫 인코딩 후 생성된 데이터프레임의 상위 5개를 확인한다.
print("\n --- 인코딩 후 --- ")
print(neighborhood_dummies.head())

 --- 인코딩 전 --- 
0    CollgCr
1    Veenker
2    CollgCr
3    Crawfor
4    NoRidge
Name: Neighborhood, dtype: object

 --- 인코딩 후 --- 
   Nbrhd_Blmngtn  Nbrhd_Blueste  Nbrhd_BrDale  Nbrhd_BrkSide  Nbrhd_ClearCr  \
0          False          False         False          False          False   
1          False          False         False          False          False   
2          False          False         False          False          False   
3          False          False         False          False          False   
4          False          False         False          False          False   

   Nbrhd_CollgCr  Nbrhd_Crawfor  Nbrhd_Edwards  Nbrhd_Gilbert  Nbrhd_IDOTRR  \
0           True          False          False          False         False   
1          False          False          False          False         False   
2           True          False          False          False         False   
3          False           True          False          False         False  

<br>

#### 📌 `원-핫 인코딩 과정 설명`: 
pd.get_dummies()를 실행하면 기존의 1개짜리 Neighborhood 컬럼이 고유한 동네 개수만큼 25개의 새로운 컬럼으로 옆으로 쫙 펼쳐진다.<br>각 집(행)마다 자신의 동네에 해당하는 컬럼에만 1 (True)을 켜고, <br>나머지 동네 컬럼에는 전부 0 (False)을 채워서 컴퓨터가 계산할 수 있는 숫자 형태로 확장을 해준 것이다.

#### 📌 `실행결과`: 1번 행의 주택은 'Veenker' 동네에 있으므로 Nbrhd_Veenker 컬럼에만 True(1)가 표시되고, 나머지는 모두 False(0)가 된다.
#### 📌 `One-Hot Encoding`: 카테고리 간의 순서나 중요도 관계를 완전히 없애므로, 모델이 잘못된 순서 관계를 학습하지 않게 한다. <br><br><br>

<div style="border: 1px solid #e0e0e0; background-color: #f8f9fa; padding: 15px; border-radius: 8px;">
    
#### `< Label Encoding 　 VS 　 One-Hot Encoding > 요약` <br>

| 구분 | 라벨 인코딩 (Label Encoding) | 원-핫 인코딩 (One-Hot Encoding) |
| :--- | :--- | :--- |
| **변환 방식** | 문자열에 <b>숫자(0, 1, 2, ...)</b>를 직접 매김 | 종류별로 <b>새 컬럼(0과 1)</b>을 여러 개 만들어 펼침 |
| **컬럼 개수** | 기존 카테고리 개수 그대로 유지 | 카테고리 개수만큼 N개로 증가 |
| **예시 (과일)** | 사과=0, 바나나=1, 딸기=2 | [과일_사과, 과일_바나나, 과일_딸기] |
</div>
<br><br><br>

<br><br><br>
# 3. `컬럼명과 타입 변경: 데이터의 뼈대 바로잡기`

#### 데이터 프레임의 구조를 직접 변경하는 세 가지 핵심 기술을 배운다.
> 1 컬럼의 이름을 알아보기 쉽게 바꾸기 　 2 각 컬럼의 데이터 타입을 목적에 맞게 변환 　 3 분석에 불필요한 컬럼은 과감하게 제거하기

<br><br>

## 🚩 1. 컬럼명 변경하기(rename)
> rename() 함수로 각 컬럼의 이름을 더 직관적으로 바꿀 수 있다. <br> <b>`'기존 이름: 새 이름'`</b> 형식의 딕셔너리(dictionary)를 사용하여 어떤 컬럼을 어떻게 바꿀지 지정한다.

In [24]:
# 원본 데이터의 컬럼 일부를 확인한다.
print(" --- 변경 전 컬럼 이름 --- ")
print(df[['GrLivArea', 'YrSold']].columns)

# .rename() 함수를 사용해 컬럼 이름을 변경한다.                           # inplace : 원본 데이터프레임(df)에 변경 사항을 즉시 직접 반영(덮어쓰기)하는 옵션
df.rename(columns={'GrLivArea' : 'LivingArea', 'YrSold' : 'Year_Sold'}, inplace=True)  
# GrLivArea(지상층 생활 면적)를 LivingArea로,  YrSold(판매 연도)를 Year_Sold로 변경한다.

# 변경 후 컬럼 이름을 다시 확인한다.
print("\n --- 변경 후 컬럼 이름 --- ")
print(df[['LivingArea', 'Year_Sold']].columns)

 --- 변경 전 컬럼 이름 --- 
Index(['GrLivArea', 'YrSold'], dtype='object')

 --- 변경 후 컬럼 이름 --- 
Index(['LivingArea', 'Year_Sold'], dtype='object')


<br><br><br>

## 🚩 2. 데이터 타입 바꾸기(astype)
> 컬럼의 성격(숫자, 문자, 날짜 등)을 분석 목적에 맞게 올바른 형식으로 바꿔주는 작업

<br>

#### ****데이터 타입을 바꾸는 경우****
> `숫자 모양의 문자열`일 때: 　 &rarr; 　 "100" 　 &rarr; 　 숫자로 보여도 타입이 object(문자)면 +,-연산이나 .mean()같은 통계 계산이 불가능하다.<br>
> `날짜 모양의 문자열`일 때: 　 &rarr; 　 "2026-04-01" 　 &rarr; 　 datetime 타입으로 바꿔야 연도/월/일 추출 및 날짜 간격 계산이 가능해진다.<br>
> `메모리를 절약`해야 할 때: 　 &rarr; 　 카테고리형 문자열을 category 타입으로 변환하면 메모리 사용량이 급격히 줄어든다.

<br>

#### ****주요 데이터 타입 종류****
| 타입명 | 판다스 표시 (`dtype`) | 설명 | 예시 |
| :--- | :--- | :--- | :--- |
| **정수형** | `int64` | 소수점이 없는 숫자 | `1`, `100` |
| **실수형** | `float64` | 소수점이 있는 숫자 | `3.14`, `12.0` |
| **문자형/범주형** | `object` / `string` | 텍스트 데이터 | `'SEOUL'`, `'A'` |
| **날짜/시간형** | `datetime64` | 날짜 및 시간 정보 | `2026-04-01` |
| **불리언** | `bool` | 참/거짓 | `True`, `False` |

In [25]:
# astype() 함수: 잘못 지정된 데이터 타입을 올바르게 변환해 준다.

# <데이터 소개> MSSubClass(주택 유형) 컬럼은 숫자로 이루어져 있지만, 실제 의미는 '1층 주택', '2층 주택' 등을 나타내는 '범주'이다.
# 모델이 2층 주택이 1층 주택보다 2배 더 중요하다는 식으로 잘못 학습할 수 있다.

# 변경 전 'MSSubClass' 컬럼의 데이터 타입을 확인한다.
print(" --- 변경 전 데이터 타입 --- ")
print(df['MSSubClass'].dtype)

# .astype()을 사용해 데이터 타입을 'object' (문자열/범주형)으로 변경한다.
df['MSSubClass'] = df['MSSubClass'].astype('object')
# astype('object')를 통해   숫자형(int64)에서 => 문자열/범주형(object)으로 변형

# 변경 후 'MSSubClass" 컬럼의 데이터 타입을 다시 확인한다.
print("\n --- 변경 후 데이터 타입 --- ")
print(df['MSSubClass'].dtype)

 --- 변경 전 데이터 타입 --- 
int64

 --- 변경 후 데이터 타입 --- 
object


<br><br><br>

## 🚩 3.불필요한 컬럼 제거하기(drop)
> 분석이나 머신러닝 모델 학습에 필요 없는 열(Column)을 선별하여 데이터프레임에서 삭제하는 작업

<br>

#### ****불필요한 컬럼을 제거하는 경우****
> `모델 성능 향상` : 　 ID, 이름 등 예측에 아무 영향을 주지 않거나 오히려 노이즈가 되는 피처(Feature)를 지워 오버피팅(과적합)을 방지한다. <br>
> `가독성 확보`<br>
> `메모리 절약 및 속도 향상`<br>

<br>

In [26]:
# <데이터 소개> Id 컬럼은 단순히 각 행을 구분하기 위한 일련변호이며 아무런 정보를 제공하지 않는다.  =>  분석에 불필요하거나 모델에 혼란을 주는 컬럼  =>  제거한다.

# drop() 함수에 columns 인자를 사용 : 특정 컬럼을 제거할 수 있다.


# 1) 제거 전 데이터프레임의 형태(차원)을 확인한다.
print("--- 제거 전 데이터 형태 ---")
print(df.shape)

# 2) .drop() 함수를 사용해 'Id'컬럼을 제거한다.
df = df.drop(columns=['Id'])        # df.drop('Id', axis=1, inplace=True) 로 적어도 된다.   =>  의미: (레이블명, 컬럼 선택, 바로 원본 교체)

# 3) 제거 후 데이터프레임의 형태를 다시 확인한다.
print("\n --- 제거 후 데이터 형태 --- ")
print(df.shape)

--- 제거 전 데이터 형태 ---
(1460, 82)

 --- 제거 후 데이터 형태 --- 
(1460, 81)


#### 📌 `실행결과`: 불필요한 Id 컬럼이 성공적으로 제거되었다.<br><br><br>

<br><br><br>
# 4. `X, Y 데이터 분리: 시험지와 정답지 분리하기`
> 머신러닝 모델을 학습시키기 위해 문제(X: 피처)와 정답(y: 타겟)을 별도의 상자로 따로 나눠담는 작업
 
### ****X와 y의 차이****
> `X (시험지 / 문제)`: 예측에 사용되는 특성(Feature) 데이터 (예: 집 크기, 방 개수, 동네 등)<br>
> `y (정답지 / 정답)`: 맞혀야 하는 목표(Target) 데이터 (예: 집값)

<br>

In [30]:
# <데이터 소개>
# House Prices 데이터에서
# SalePrice                          =>   우리가 예측해야 할 '정답' 이므로   =>   y
# SalePrice를 제외한 나머지 모든 컬럼   =>   정답을 예측하기 위한 '문제' 이므로 =>   X


# 1) 'SalePrice'컬럼을 선택하여 y (Target) 변수에 저장한다.
y = df['SalePrice']

# 2) 원본 데이터프레임 df에서 'SalePrice'컬럼을 제거하여 X(Feature) 변수에 저장한다.
X = df.drop(columns=["SalePrice"])

# 3) 분리된 X와 y 데이터의 형태(shape)를 각각 출력하여 확인한다.
print("\n --- X (Feature) 데이터 형태 ---")
print(X.shape)

print("\n --- y (Target) 데이터 형태 ---")
print(y.shape)


 --- X (Feature) 데이터 형태 ---
(1460, 80)

 --- y (Target) 데이터 형태 ---
(1460,)


#### 📌 `실행결과`: X 데이터는 1460개의 행과 80개의 열을, y 데이터는 1460개의 값을 가진 단일 열(Series)임을 알 수 있다.<br><br><br>

<br><br><br>
# 5. `데이터 결합(merge)`
> 공통된 기준 컬럼(Key)을 바탕으로 서로 다른 두 데이터프레임을 하나로 엮는 작업

📌실무에서는 모든 정보가 하나의 표에 들어있지 않다.<br>
EX) '고객 정보 표'와 '주문 내역 표'가 따로 있을 때, <br>
두 표에 공통으로 들어있는 '고객 ID'를 기준으로 합쳐야 "어떤 고객이 무슨 상품을 샀는지" 한눈에 분석할 수 있다.

<br>

## 🚩 1. merge의 핵심 원리: 공통된 열쇠(key)로 연결하기
> 두 개의 서로 다른 데이터프레임을 특정 '기준(Key)'이 되는 열을 중심으로 결합한다.

In [4]:
# < 데이터 소개 >
# 데이터 1: 학교의 학생 명단 (학번, 이름)
# 데이터 2: 학교의 동아리 명단 (학번, 동아리명)

# ==> 모두 '학번' 이라는 공통된 정보가 있음. ==> 기준(Key)


# ============================== 코드 예제  ==============================
# 1) 학생 정보를 담은 첫 번째 데이터프레임 생성

df_students = pd.DataFrame({
    'student_id' : [1, 2, 3, 4],
    'name' : ['Alice', 'Bob', 'Charlie', 'David']
})
# 데이터는 파이썬의 Dictionary 형태로 전달한다.
# Dictionary의 Key :    'student_id', 'name'은 데이터프레임의 Column 이름이 된다.
# Dictionary의 Value :  [1, 2, 3, 4]는 해당 Column을 채우는 데이터가 된다.



# 2) 동아리 정보를 담은 두 번째 데이터프레임 생성

df_clubs = pd.DataFrame({
    'student_id' : [3, 1, 2, 5],
    'club' : ['Art', 'Science', 'Dance', 'Music']
})

print("--- 학생 명단 (df_students) ---")
print(df_students)
print("\n--- 동아리 명단 (df_clubs) ---")
print(df_clubs)

--- 학생 명단 (df_students) ---
   student_id     name
0           1    Alice
1           2      Bob
2           3  Charlie
3           4    David

--- 동아리 명단 (df_clubs) ---
   student_id     club
0           3      Art
1           1  Science
2           2    Dance
3           5    Music


<br><br><br>
## 🚩 2. pd.merge()로 데이터 합치기
> pd.merge() 함수: 결합할 두 데이터프레임과 기준이 될 Key 컬럼을 지정하여 사용

 <br>

<details>
<summary><b>🔍 pd.merge 결합의 4가지 방식 (how=옵션)</b></summary>
<br>
    
how 옵션에 따라 어떤 데이터를 남기고 어떤 데이터를 버릴지 결정된다.<br>
| 결합 방식 | `how=` 값 | 벤다이어그램 개념 | 설명 |
| :--- | :--- | :--- | :--- |
| **교집합 (Inner)** | `'inner'` | **A $\cap$ B** | 두 데이터에 **모두 존재**하는 기준값만 남김 (기본값) |
| **합집합 (Outer)** | `'outer'` | **A $\cup$ B** | 한쪽에만 존재해도 **전부 남김** (없는 값은 NaN) |
| **왼쪽 기준 (Left)** | `'left'` | **A** | **왼쪽** 데이터는 다 남기고, 오른쪽 데이터 붙임 |
| **오른쪽 기준 (Right)**| `'right'` | **B** | **오른쪽** 데이터는 다 남기고, 왼쪽 데이터 붙임 |

</details>
└ 가장 기본적인 방식은, 두 데이터프레임에 모두 존재하는 데이터만 합치는 inner 방식이다.
<br><br>

In [5]:
# pd.merge()를 사용해 두 데이터프레임을 합칩니다.
merged_df = pd.merge(df_students, df_clubs, on='student_id', how='inner')  # student_id를 기준으로 df_students와 df_clubs를 합쳐라
                                                            #how='inner' : 두 명단에 모두 'student_id'가 존재하는 학생들만 결과에 포함시켜라.

# 합쳐진 결과 데이터프레임을 출력한다.
print("--- 병합된 명단 (merged_df) ---")
print(merged_df)

--- 병합된 명단 (merged_df) ---
   student_id     name     club
0           1    Alice  Science
1           2      Bob    Dance
2           3  Charlie      Art


#### 📌 `실행결과`: 
student_id를 기준으로 name과 club 정보가 올바르게 연결됐다.<br>
student_id가 1인 'Alice'는 'Science'동아리에, 2인'Bob'은 'Dance'동아리에 소속되어 있다.<br>
<b>`'학생 명단'에만 있던 'David'(ID 4)<br>와 '동아리 명단'에만 있던 'Music'동아리원(ID 5)는 최종 결과에 포함되지 않았다.`</b> 　&rarr; 　<b>`inner 방식의 특징`</b><br><br><br>